pkg/endpoint/bpf.go

```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+		e.getLogger().WithField(logfields.BPFHeaderfileHash, datapathRegenCtx.bpfHeaderfilesHash).
 			Debugf("BPF header file hashed (was: %q)", e.bpfHeaderfileHash)
 	}
 
-	// Cache endpoint information
-	// TODO (ianvernon): why do we need to do this?
-	var epInfoCache *epInfoCache
-	if bpfHeaderfilesChanged {
-		epInfoCache = e.createEpInfoCache(nextDir)
+	// Cache endpoint information so that we can release the endpoint lock
+	// but still access information about the endpoint later if needed.
+	if datapathRegenCtx.bpfHeaderfilesChanged {
+		datapathRegenCtx.epInfoCache = e.createEpInfoCache(nextDir)
 	} else {
-		epInfoCache = e.createEpInfoCache(currentDir)
+		datapathRegenCtx.epInfoCache = e.createEpInfoCache(currentDir)
 	}
-	if epInfoCache == nil {
+	if datapathRegenCtx.epInfoCache == nil {
 		stats.prepareBuild.End(false)
-		e.Unlock()
-		err = fmt.Errorf("Unable to cache endpoint information")
-		return 0, compilationExecuted, err
+		return proxyCleanupFunc, fmt.Errorf("Unable to cache endpoint information")
 	}
-
 	// TODO: In Cilium v1.4 or later cycle, remove this.
 	os.RemoveAll(e.IPv6EgressMapPathLocked())
 	os.RemoveAll(e.IPv4EgressMapPathLocked())
 	os.RemoveAll(e.IPv6IngressMapPathLocked())
 	os.RemoveAll(e.IPv4IngressMapPathLocked())
+	return proxyCleanupFunc, nil
+}
+
+// regenerateBPF rewrites all headers and updates all BPF maps to reflect the
+// specified endpoint.
+// Must be called with endpoint.Mutex not held and endpoint.BuildMutex held.
+// Returns the policy revision number when the regeneration has called, a
+// boolean if the BPF compilation was executed and an error in case of an error.
+func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenContext *RegenerationContext) (revnum uint64, compiled bool, reterr error) {
+	stats := &regenContext.Stats
+	stats.waitingForLock.Start()
+
+	// Make sure that owner is not compiling base programs while we are
+	// regenerating an endpoint.
+	owner.GetCompilationLock().RLock()
+	defer owner.GetCompilationLock().RUnlock()
+
+	// Cache endpoint ID for use later.
+	e.UnconditionalRLock()
+	epID := e.StringID()
+	e.RUnlock()
+
+	bpfRegenContext := newDatapathRegenerationContext()
+	defer bpfRegenContext.completionCancel()
 
-	e.Unlock()
+	// Wrap logic which requires the endpoint to be locked within a function
+	// & call within regenerateBPF so that endpoint lock release can be deferred
+	// within the function. If this isn't done, and something else panics
+	// while the endpoint regeneration lock is held, and a deferred function
+	// within regenerateBPF tries to grab the endpoint lock, it will deadlock
+	// because the lock is already held, which obfuscates what happened in the
+	// initial panic.
+	// Returns an error, as well as a function which may need deferring
+	// based once this function returns. This function is intentionally not
+	// deferred in this function itself because it may acquire the endpoint
+	// lock.
```

Comment:
Is the last sentence correct? I thought `revFunc` needs to be returned from `runCompilationSteps` and deferred in `regenerateBPF` because it only reverted if there was an error in either `runCompilationSteps` or `regenerateBPF` (the lock stuff also applies).
With that in mind perhaps it would be simpler to not check the error in `revFunc` but run it in `runCompilationSteps` if that returns an error, and separately run it here if regenerateBPF encounters an error. This might be worse, though, since we will likely do a `defer func() { if err != nil { proxyCleanup/revFunc() } }()` in each function anyway and that's a little more obtuse.

Reply:
I've removed this comment / returning a function to defer in the caller.